### PR TITLE
[MM-25314] Migrate AppError from channel_member_history_store.go

### LINKED ISSUE #14598: Migrate AppError to error from store/sqlstore/channel_member_history_store.go
Mattermost is migrating all usages of the AppError type to the idiomatic error type. AppError is a loaded error type containing fields like HTTP Status code which is not really necessary for errors coming from the DB layer. We should migrate to using normal types satisfying the error interface.

This ticket is about migrating all usages of AppError from `store/sqlstore/channel_member_history_store.go` to return a normal error. The migration should use the error types defined in store/errors.go. If the need should arise to add another specific error type, feel free to add it. And then in the app layer, we would need to convert the `error` type to the usual `AppError`. This means we need a switch case to check for the type of error and appropriately convert to the correct AppError. It can be something like:

```go
if _, err := a.Srv().Store.Bot().Update(bot); err != nil {
	var nfErr *store.ErrNotFound
	var appErr *model.AppError
	switch {
	case errors.As(err, &nfErr):
		return model.MakeBotNotFoundError(nfErr.Id)
	case errors.As(err, &appErr): // in case we haven't converted to plain error.
		return appErr
	default: // last fallback in case it doesn't map to an existing app error.
		return model.NewAppError("DeleteBotIconImage", "app.bot.patchbot.internal_error", nil, err.Error(), http.StatusInternalServerError)
	}
}
```

Example Pull Request: #14339

If you have questions about the ticket or you need any help, feel free to contact `agniva.de.sarker` in https://community.mattermost.com/

----------------------------------

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-25314

### PR DESCRIPTION
**Pull request by dantepippi**
**Summary**

This PR migrates channel_member_history_store.go AppError type usages to the idiomatic error type.

JIRA Ticket: https://mattermost.atlassian.net/browse/MM-25314

Closes #14598

### CODE REVIEW

File: store/sqlstore/channel_member_history_store.go [hunk 1]
```diff
@@ -5,11 +5,11 @@ package sqlstore
 
 import (
 	"database/sql"
-	"net/http"
 
 	"github.com/mattermost/mattermost-server/v5/mlog"
 	"github.com/mattermost/mattermost-server/v5/model"
 	"github.com/mattermost/mattermost-server/v5/store"
+	"github.com/pkg/errors"
```
Comment (agnivade) [inline store/sqlstore/channel_member_history_store.go:12]: Nit: Let's place this in a separate block

File: store/sqlstore/channel_member_history_store.go [hunk 2]
```diff
@@ -181,12 +180,12 @@ func (s SqlChannelMemberHistoryStore) PermanentDeleteBatch(endTime int64, limit
 	params := map[string]interface{}{"EndTime": endTime, "Limit": limit}
 	sqlResult, err := s.GetMaster().Exec(query, params)
 	if err != nil {
-		return int64(0), model.NewAppError("SqlChannelMemberHistoryStore.PermanentDeleteBatchForChannel", "store.sql_channel_member_history.permanent_delete_batch.app_error", params, err.Error(), http.StatusInternalServerError)
+		return 0, errors.Wrapf(err, "PermanentDeleteBatchForChannel endTime=%d limit=%d", endTime, limit)
```
Comment (agnivade) [inline store/sqlstore/channel_member_history_store.go:183]: PermanentDeleteBatch

File: store/sqlstore/channel_member_history_store.go [hunk 3]
```diff
@@ -181,12 +180,12 @@ func (s SqlChannelMemberHistoryStore) PermanentDeleteBatch(endTime int64, limit
...
 	}
 
 	rowsAffected, err := sqlResult.RowsAffected()
 	if err != nil {
-		return int64(0), model.NewAppError("SqlChannelMemberHistoryStore.PermanentDeleteBatchForChannel", "store.sql_channel_member_history.permanent_delete_batch.app_error", params, err.Error(), http.StatusInternalServerError)
+		return 0, errors.Wrapf(err, "PermanentDeleteBatchForChannel endTime=%d limit=%d", endTime, limit)
```
Comment (agnivade) [inline store/sqlstore/channel_member_history_store.go:188]: Same as above

File: store/sqlstore/channel_member_history_store.go [hunk 4]
```diff
@@ -65,18 +65,18 @@ func (s SqlChannelMemberHistoryStore) LogLeaveEvent(userId string, channelId str
 	return nil
 }
 
-func (s SqlChannelMemberHistoryStore) GetUsersInChannelDuring(startTime int64, endTime int64, channelId string) ([]*model.ChannelMemberHistoryResult, *model.AppError) {
+func (s SqlChannelMemberHistoryStore) GetUsersInChannelDuring(startTime int64, endTime int64, channelId string) ([]*model.ChannelMemberHistoryResult, error) {
 	useChannelMemberHistory, err := s.hasDataAtOrBefore(startTime)
 	if err != nil {
-		return nil, model.NewAppError("SqlChannelMemberHistoryStore.GetUsersInChannelAt", "store.sql_channel_member_history.get_users_in_channel_during.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return nil, errors.Wrapf(err, "GetUsersInChannelDuring startTime=%d endTime=%d channelId=%s", startTime, endTime, channelId)
```
Comment (agnivade) [inline store/sqlstore/channel_member_history_store.go:71]: hasDataAtOrBefore

File: store/sqlstore/channel_member_history_store.go [hunk 5]
```diff
@@ -86,7 +86,7 @@ func (s SqlChannelMemberHistoryStore) GetUsersInChannelDuring(startTime int64, e
 	// this may not always be true, but it's better than saying that somebody wasn't there when they were
 	channelMemberHistories, err := s.getFromChannelMembersTable(startTime, endTime, channelId)
 	if err != nil {
-		return nil, model.NewAppError("SqlChannelMemberHistoryStore.GetUsersInChannelAt", "store.sql_channel_member_history.get_users_in_channel_during.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return nil, errors.Wrapf(err, "GetUsersInChannelAt startTime=%d endTime=%d channelId=%s", startTime, endTime, channelId)
```
Comment (agnivade) [inline store/sqlstore/channel_member_history_store.go:89]: getFromChannelMembersTable

File: app/channel.go [hunk 6]
```diff
@@ -97,7 +97,13 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
 		_, err = a.Srv().Store.Channel().SaveMember(cm)
 		if histErr := a.Srv().Store.ChannelMemberHistory().LogJoinEvent(user.Id, channel.Id, model.GetMillis()); histErr != nil {
 			mlog.Error("Failed to update ChannelMemberHistory table", mlog.Err(histErr))
-			return histErr
+			var appErr *model.AppError
```
Comment (agnivade) [inline app/channel.go:100]: I don't see any case where we are actually returning an AppError. So in that case I think we can directly return an AppError without any switch case. Same applies for all of these below.

File: store/sqlstore/channel_member_history_store.go (hunk 2 above)
Comment (dantepippi) [reply store/sqlstore/channel_member_history_store.go:183]: My bad! Good catch!

File: app/channel.go (hunk 6 above)
Comment (dantepippi) [reply app/channel.go:100]: Great, that cleaned a lot.

**Comment by jespino** [comment]
I merged the master branch and updated the i18n file properly.

**Comment by lindy65** [comment]
Hi @agnivade - can you help with test steps for this PR please?  Is testing creating all types of channels, adding and removing members the way to go?

**Comment by agnivade** [comment]
Yep, that's basically it. Create DM/GM//Public/Private channels, add/remove users from them. Spot check for any issues. And done. :+1:

**Comment by lindy65** [comment]
Awesome, thanks for confirming @agnivade 👍

**Comment by agnivade** [comment]
Well, this is a bummer. The EE codebase has some references to this store which needs to be fixed before this goes in. To make it work, we'd need to have the same branch name in both EE and TE repos. And since you don't have upstream write access, you can't create a branch directly.

I'll close this and create another PR from your commits in a new branch and fix the EE issues as well. Sorry for the troubles @dantepippi.

**Comment by agnivade** [comment]
Created new PR here https://github.com/mattermost/mattermost-server/pull/14693.

**Comment by dantepippi** [comment]
> Well, this is a bummer. The EE codebase has some references to this store which needs to be fixed before this goes in. To make it work, we'd need to have the same branch name in both EE and TE repos. And since you don't have upstream write access, you can't create a branch directly.
> 
> I'll close this and create another PR from your commits in a new branch and fix the EE issues as well. Sorry for the troubles @dantepippi.

No worries @agnivade . Hope I was able to help here.

**Pull request closed without merging**